text/3735-field-projections.md

```diff
@@ -0,0 +1,1555 @@
+- `fut->fut1: Pin<&mut F1>`
+- `fut->fut2: Pin<&mut F2>`
+- `fut->fair: &mut bool`
+
+Using these, one can concisely implement `Future` for `FairRaceFuture` without any `unsafe` code:
+
+```rust
+impl<F1: Future, F2: Future<Output = F1::Output>> Future for FairRaceFuture<F1, F2> {
+    type Output = F1::Output;
+
+    fn poll(self: Pin<&mut Self>, ctx: &mut Context) -> Poll<Self::Output> {
+        let fair: &mut bool = self->fair;
+        *fair = !*fair;
+        if *fair {
+            // self->fut1: Pin<&mut F1>
+            match self->fut1.poll(ctx) {
+                Poll::Pending => self->fut2.poll(ctx),
+                Poll::Ready(res) => Poll::Ready(res),
+            }
+        } else {
+            // self->fut2: Pin<&mut F2>
+            match self->fut2.poll(ctx) {
+                Poll::Pending => self->fut1.poll(ctx),
+                Poll::Ready(res) => Poll::Ready(res),
+            }
+        }
+    }
+}
+```
+
+## Impact of this Feature
+
+Overall this feature improves readability of code, because it replaces more complex to parse syntax
+with simpler syntax:
+- `&raw mut (*ptr).foo` is turned into `ptr->foo`
+- using `NonNull<T>` as a replacement for `*mut T` becomes a lot better when accessing fields,
+
+There is of course a cost associated with introducing a new operator along with the concept of field
+projections.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Implementation Details
+
+In order to facilitate field projections, several interlinked concepts have to be introduced. These
+concepts are:
+
+- [field types],
+  - `Field` trait,
+  - `field_of!` macro,
+  - [`#[projecting]`](#projecting-attribute) attribute
+- projection operator `->`,
+  - `Project` trait,
+  - `Projectable` trait,
+
+To ease understanding, here is a short explanation of the interactions between these concepts. The
+following subsections explain them in more detail, so refer to them in cases of ambiguity. The
+projection operator `->` is governed by the `Project` trait that has `Projectable` as a super trait.
+`Projectable` helps to select the struct whose fields are used for projection. Field types store
+information about a field (such as the base struct and the field type) via the `Field` trait and the
+`field_of!` macro makes it possible to name the [field types]. Finally, the
+[`#[projecting]`](#projecting-attribute) attribute allows `repr(transparent)` structs to be ignored
+when looking for fields for projection.
+
+### Field Types
+[field type]: #field-types
+[field types]: #field-types
+
+The compiler generates a compiler-internal type for every field of every struct. These types can
+only be named via the `field_of!` macro that has the same syntax as `offset_of!`. Only fields
+accessible to the current scope can be projected. These types are called *field types*.
+
+Field types are not generated for fields of `#[repr(packed)]` structs that are misaligned.
```

Comment:
Actually this may be too restrictive. It is safe and sound to create a raw pointer to a misaligned field:

```rust
#[repr(C, packed)]
struct X {
    a: u8,
    b: u16, // b is misaligned
}

fn main() {
    assert_eq!(offset_of!(X, b), 1); // offset_of works
    let x = X { a: 1, b: 2 };
    let px = &raw const x;
    let pb = &raw const x.b; // works and safe, unlike `&x.b` which is E0793
//  let pb = unsafe { &raw const (*px).b }; // equivalent to above, unsafe{} needed for dereference
    assert_eq!(unsafe { pb.byte_sub_ptr(px) }, 1);
}
```
this suggested `let pb = px->b;` should be valid.

I think this could be supported by changing the Field trait to 
```rust
pub unsafe trait UnalignedField {
    type Base: ?Sized;
    type Type: ?Sized;
    const OFFSET: usize;
}
pub unsafe trait Field: UnalignedField {}
```
Containers supporting unaligned projection (`*T` and `NonNull<T>`) can use `F: UnalignedField` as the bound when implementing `Project<F>`.

Reply:
done :+1: